### ISSUE TITLE
[Question] Using two action terms in the PreTrainedPolicyAction used in the Manager based Navigation Example

### ISSUE THREAD

**Issue by Parkersep** [isaac-sim/IsaacLab#2910]
I am having some difficulty figuring out the best way to combine the velocity and position commands I used for my locomotion policy into PreTrainedPolicyAction used in the manager navigation example. Could I call all of the action terms used in ActionCfg instead of only being able to choose one?  Would adding a way to pass action terms then combining the action tensors be the way to go about that? Thanks for any help!

`
class PreTrainedPolicyAction(ActionTerm):
 

    cfg: PreTrainedPolicyActionCfg
    """The configuration of the action term."""

    def __init__(self, cfg: PreTrainedPolicyActionCfg, env: ManagerBasedRLEnv) -> None:
        # initialize the action term
        super().__init__(cfg, env)

        self.robot: Articulation = env.scene[cfg.asset_name]

        # load policy
        if not check_file_path(cfg.policy_path):
            raise FileNotFoundError(f"Policy file '{cfg.policy_path}' does not exist.")
        file_bytes = read_file(cfg.policy_path)
        self.policy = torch.jit.load(file_bytes).to(env.device).eval()

        self._raw_actions = torch.zeros(self.num_envs, self.action_dim, device=self.device)

        # prepare low level actions
        self._low_level_action_term: ActionTerm = cfg.low_level_actions.class_type(cfg.low_level_actions, env)
        self.low_level_actions = torch.zeros(self.num_envs, self._low_level_action_term.action_dim, device=self.device)

        def last_action():
            # reset the low level actions if the episode was reset
            if hasattr(env, "episode_length_buf"):
                self.low_level_actions[env.episode_length_buf == 0, :] = 0
            return self.low_level_actions

        # remap some of the low level observations to internal observations
        cfg.low_level_observations.actions.func = lambda dummy_env: last_action()
        cfg.low_level_observations.actions.params = dict()
        cfg.low_level_observations.velocity_commands.func = lambda dummy_env: self._raw_actions
        cfg.low_level_observations.velocity_commands.params = dict()

        # add the low level observations to the observation manager
        self._low_level_obs_manager = ObservationManager({"ll_policy": cfg.low_level_observations}, env)

        self._counter = 0

    """
    Properties.
    """

    @property
    def action_dim(self) -> int:
        return 3

    @property
    def raw_actions(self) -> torch.Tensor:
        return self._raw_actions

    @property
    def processed_actions(self) -> torch.Tensor:
        return self.raw_actions

    """
    Operations.
    """

    def process_actions(self, actions: torch.Tensor):
        self._raw_actions[:] = actions

    def apply_actions(self):
        if self._counter % self.cfg.low_level_decimation == 0:
            low_level_obs = self._low_level_obs_manager.compute_group("ll_policy")
            self.low_level_actions[:] = self.policy(low_level_obs)
            self._low_level_action_term.process_actions(self.low_level_actions)
            self._counter = 0
        self._low_level_action_term.apply_actions()
        self._counter += 1

    """
    Debug visualization.
    """

    def _set_debug_vis_impl(self, debug_vis: bool):
        # set visibility of markers
        # note: parent only deals with callbacks. not their visibility
        if debug_vis:
            # create markers if necessary for the first tome
            if not hasattr(self, "base_vel_goal_visualizer"):
                # -- goal
                marker_cfg = GREEN_ARROW_X_MARKER_CFG.copy()
                marker_cfg.prim_path = "/Visuals/Actions/velocity_goal"
                marker_cfg.markers["arrow"].scale = (0.5, 0.5, 0.5)
                self.base_vel_goal_visualizer = VisualizationMarkers(marker_cfg)
                # -- current
                marker_cfg = BLUE_ARROW_X_MARKER_CFG.copy()
                marker_cfg.prim_path = "/Visuals/Actions/velocity_current"
                marker_cfg.markers["arrow"].scale = (0.5, 0.5, 0.5)
                self.base_vel_visualizer = VisualizationMarkers(marker_cfg)
            # set their visibility to true
            self.base_vel_goal_visualizer.set_visibility(True)
            self.base_vel_visualizer.set_visibility(True)
        else:
            if hasattr(self, "base_vel_goal_visualizer"):
                self.base_vel_goal_visualizer.set_visibility(False)
                self.base_vel_visualizer.set_visibility(False)

    def _debug_vis_callback(self, event):
        # check if robot is initialized
        # note: this is needed in-case the robot is de-initialized. we can't access the data
        if not self.robot.is_initialized:
            return
        # get marker location
        # -- base state
        base_pos_w = self.robot.data.root_pos_w.clone()
        base_pos_w[:, 2] += 0.5
        # -- resolve the scales and quaternions
        vel_des_arrow_scale, vel_des_arrow_quat = self._resolve_xy_velocity_to_arrow(self.raw_actions[:, :2])
        vel_arrow_scale, vel_arrow_quat = self._resolve_xy_velocity_to_arrow(self.robot.data.root_lin_vel_b[:, :2])
        # display markers
        self.base_vel_goal_visualizer.visualize(base_pos_w, vel_des_arrow_quat, vel_des_arrow_scale)
        self.base_vel_visualizer.visualize(base_pos_w, vel_arrow_quat, vel_arrow_scale)

    """
    Internal helpers.
    """

    def _resolve_xy_velocity_to_arrow(self, xy_velocity: torch.Tensor) -> tuple[torch.Tensor, torch.Tensor]:
        """Converts the XY base velocity command to arrow direction rotation."""
        # obtain default scale of the marker
        default_scale = self.base_vel_goal_visualizer.cfg.markers["arrow"].scale
        # arrow-scale
        arrow_scale = torch.tensor(default_scale, device=self.device).repeat(xy_velocity.shape[0], 1)
        arrow_scale[:, 0] *= torch.linalg.norm(xy_velocity, dim=1) * 3.0
        # arrow-direction
        heading_angle = torch.atan2(xy_velocity[:, 1], xy_velocity[:, 0])
        zeros = torch.zeros_like(heading_angle)
        arrow_quat = math_utils.quat_from_euler_xyz(zeros, zeros, heading_angle)
        # convert everything back from base to world frame
        base_quat_w = self.robot.data.root_quat_w
        arrow_quat = math_utils.quat_mul(base_quat_w, arrow_quat)

        return arrow_scale, arrow_quat
'

**Comment by RandomOakForest** [comment]
Thank you for posting this. If you closed this by accident, feel free to reopen it. If you resolved the issue, it would be great if you could share your solution. Thanks.

**Comment by Parkersep** [comment]
Was able to find a fix . I altered the PreTrainedPolicyclass so I could pass in two action terms and then apply the policy outptuts to the correct joint groups.

Here's the updated code for reference. The only issue I have now is when running with the headless flag I get no errors, but when I dont have the flag I get a error with a tensor being sent to the CPU. Not a huge issue since when I play the trained policy I can just use the cpu. Just wanted to mention that in case anyone else come across thie same issue. 


```
from __future__ import annotations

from dataclasses import MISSING
from typing import TYPE_CHECKING
import torch

import isaaclab.utils.math as math_utils
from isaaclab.assets import Articulation
from isaaclab.managers import (
    ActionTerm,
    ActionTermCfg,
    ObservationGroupCfg,
    ObservationManager,
)
from isaaclab.markers import VisualizationMarkers
from isaaclab.markers.config import BLUE_ARROW_X_MARKER_CFG, GREEN_ARROW_X_MARKER_CFG
from isaaclab.utils import configclass
from isaaclab.utils.assets import check_file_path, read_file

if TYPE_CHECKING:
    from isaaclab.envs import ManagerBasedRLEnv


class PreTrainedPolicyAction(ActionTerm):
    r"""Runs a TorchScript policy and dispatches its 16-D output into two
    low-level ActionTerms:
      • first `pos_dim` values → term #1  (e.g. joint positions)
      • last  `vel_dim` values → term #2  (e.g. wheel velocities)
    """

    cfg: "PreTrainedPolicyActionCfg"  # for type-checkers

    def __init__(self, cfg: "PreTrainedPolicyfrom __future__ import annotations

from dataclasses import MISSING
from typing import TYPE_CHECKING
import torchActionCfg", env: "ManagerBasedRLEnv"):
        super().__init__(cfg, env)

        self.robot: Articulation = env.scene[cfg.asset_name]

        if not check_file_path(cfg.policy_path):
            raise FileNotFoundError(f"Policy file '{cfg.policy_path}' does not exist.")
        file_bytes = read_file(cfg.policy_path)
        self.policy = torch.jit.load(file_bytes).to(env.device).eval()

        self._raw_actions = torch.zeros(
            self.num_envs, self.action_dim, device=self.device
        )

        self._ll_term_1: ActionTerm = cfg.low_level_actions_pos.class_type(
            cfg.low_level_actions_pos, env
        )
        self._ll_term_2: ActionTerm = cfg.low_level_actions_vel.class_type(
            cfg.low_level_actions_vel, env
        )

        self.pos_dim = self._ll_term_1.action_dim  # usually 12
        self.vel_dim = self._ll_term_2.action_dim  # usually 4
        self.legs_slice = slice(0, self.pos_dim)  # [:12]
        self.wheels_slice = slice(
            self.pos_dim,
            self.pos_dim + self.vel_dim,
        )  # [12:16]

        self.low_level_actions = torch.zeros(  # (B,16)
            self.num_envs, self.pos_dim + self.vel_dim, device=self.device
        )

        def last_action():
            if hasattr(env, "episode_length_buf"):
                self.low_level_actions[env.episode_length_buf == 0, :] = 0
            return self.low_level_actions

        cfg.low_level_observations.actions.func = lambda _env: last_action()
        cfg.low_level_observations.actions.params = dict()
        cfg.low_level_observations.velocity_commands.func = (
            lambda _env: self._raw_actions
        )
        cfg.low_level_observations.velocity_commands.params = dict()
        self._ll_obs_mgr = ObservationManager(
            {"ll_policy": cfg.low_level_observations}, env
        )

        self._counter = 0

    @property
    def action_dim(self) -> int:
        return 3

    @property
    def raw_actions(self) -> torch.Tensor:
        return self._raw_actions

    @property
    def processed_actions(self) -> torch.Tensor:
        return self._raw_actions

    def process_actions(self, actions: torch.Tensor):
        self._raw_actions[:] = actions

    def apply_actions(self):
        if self._counter % self.cfg.low_level_decimation == 0:
            ll_obs = self._ll_obs_mgr.compute_group("ll_policy")
            if ll_obs.device != self.device:
                ll_obs = ll_obs.to(self.device)
            full_out = self.policy(ll_obs)
            self.low_level_actions[:] = full_out

            self._ll_term_1.process_actions(full_out[:, self.legs_slice])
            self._ll_term_2.process_actions(full_out[:, self.wheels_slice])

            self._counter = 0

        self._ll_term_1.apply_actions()
        self._ll_term_2.apply_actions()
        self._counter += 1


@configclass
class PreTrainedPolicyActionCfg(ActionTermCfg):
    """Config for :class:`PreTrainedPolicyAction`."""

    class_type: type[ActionTerm] = PreTrainedPolicyAction
    asset_name: str = MISSING

    policy_path: str = MISSING  # TorchScript (.pt)
    low_level_decimation: int = 4

    low_level_actions_pos: ActionTermCfg = MISSING  # legs (pos)
    low_level_actions_vel: ActionTermCfg = MISSING  # wheels (vel)
    low_level_observations: ObservationGroupCfg = MISSING

    debug_vis: bool = False
```
